File: src/ilink.ts
/**
 * iLink API — WeChat bot gateway
 * Protocol lifted from weixin-claude-code (MIT), rewritten for clarity.
 *
 * Ownership: this module owns no persistent state.
 * Callers are responsible for storing sync_buf between calls.
 */

import fs from "node:fs";
import path from "node:path";
import os from "node:os";
import crypto from "node:crypto";
import { LONG_POLL_TIMEOUT_MS, API_TIMEOUT_MS } from "./config.js";

export const CRED_PATH = path.join(os.homedir(), ".claude", "channels", "wechat", "account.json");
export const SYNC_PATH = path.join(os.homedir(), ".claude", "channels", "wechat", "sync_buf.txt");

export interface ILinkCreds {
  token: string;
  baseUrl: string;
}

export interface ILinkMessage {
  msg_type: number;       // 1 = user, 2 = bot
  from_user: string;
  context_token: string;
  content?: { type: number; text?: string }[];
}

export interface GetUpdatesResponse {
  ret?: number;
  errcode?: number;
  msgs?: ILinkMessage[];
  get_updates_buf?: string;
}

// ── Credentials ─────────────────────────────────────────

export function loadCreds(): ILinkCreds {
  if (!fs.existsSync(CRED_PATH)) {
    throw new Error(`No WeChat credentials. Run: npm run setup`);
  }
  const raw = JSON.parse(fs.readFileSync(CRED_PATH, "utf-8"));
  if (!raw.token) throw new Error("WeChat credentials missing token field");
  return { token: raw.token, baseUrl: raw.baseUrl || "https://ilinkai.weixin.qq.com" };
}

export function loadSyncBuf(): string {
  try { return fs.existsSync(SYNC_PATH) ? fs.readFileSync(SYNC_PATH, "utf-8") : ""; }
  catch { return ""; }
}

export function saveSyncBuf(buf: string): void {
  fs.mkdirSync(path.dirname(SYNC_PATH), { recursive: true });
  fs.writeFileSync(SYNC_PATH, buf, "utf-8");
}

// ── HTTP helpers ─────────────────────────────────────────

function buildHeaders(token: string, body?: string): Record<string, string> {
  const h: Record<string, string> = {
    "Content-Type":    "application/json",
    Authorization:     `Bearer ${token}`,
    AuthorizationType: "ilink_bot_token",
    "X-WECHAT-UIN":   Buffer.from(String(crypto.randomBytes(4).readUInt32BE(0))).toString("base64"),
  };
  if (body) h["Content-Length"] = String(Buffer.byteLength(body, "utf-8"));
  return h;
}

// ── API calls ────────────────────────────────────────────

export async function getUpdates(
  creds: ILinkCreds,
  syncBuf: string,
): Promise<GetUpdatesResponse> {
  const body = JSON.stringify({ get_updates_buf: syncBuf });
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), LONG_POLL_TIMEOUT_MS);
  try {
    const res = await fetch(`${creds.baseUrl}/ilink/bot/getupdates`, {
      method: "POST",
      headers: buildHeaders(creds.token, body),
      body,
      signal: ctrl.signal,
    });
    clearTimeout(timer);
    return await res.json() as GetUpdatesResponse;
  } catch (err: unknown) {
    clearTimeout(timer);
    if (err instanceof Error && err.name === "AbortError") {
      return { ret: 0, msgs: [], get_updates_buf: syncBuf };
    }
    throw err;
  }
}

export async function sendReply(
  creds: ILinkCreds,
  toUser: string,
  text: string,
  contextToken: string,
): Promise<void> {
  const payload = {
    to_user: toUser,
    context_token: contextToken,
    msg_items: [{ item_type: 1, text_content: { content: text } }],
  };
  const body = JSON.stringify(payload);
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), API_TIMEOUT_MS);
  try {
    await fetch(`${creds.baseUrl}/ilink/bot/sendmessage`, {
      method: "POST",
      headers: buildHeaders(creds.token, body),
      body,
      signal: ctrl.signal,
    });
  } finally {
    clearTimeout(timer);
  }
}

export async function getBotQRCode(baseUrl: string): Promise<{ qr_url: string; ticket: string; token: string }> {
  const body = JSON.stringify({ bot_type: 3 });
  const res = await fetch(`${baseUrl}/ilink/bot/get_bot_qrcode`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body,
  });
  return await res.json() as { qr_url: string; ticket: string; token: string };
}

export async function pollQRLogin(
  baseUrl: string,
  ticket: string,
): Promise<{ status: number; token?: string }> {
  const body = JSON.stringify({ ticket });
  const res = await fetch(`${baseUrl}/ilink/bot/query_qrcode_status`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body,
  });
  return await res.json() as { status: number; token?: string };
}

File: cpp/src/ilink.cpp
// wechat-copilot — RustCC C++17 edition
// ilink.cpp — WeChat iLink gateway implementation
//
// TCC-OWN:  All heap strings returned by value (move).
// TCC-LIFE: json objects are temporaries; fields extracted before scope ends.
// TCC-CONC: All functions are stateless; thread-safe by design.

#include "wcp/ilink.hpp"
#include "wcp/http.hpp"

#include <nlohmann/json.hpp>

#include <chrono>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <random>
#include <sstream>
#include <stdexcept>

namespace wcp {

using json = nlohmann::json;
namespace fs = std::filesystem;

// ── helpers ────────────────────────────────────────────────────────────────

namespace {

std::string random_hex(size_t bytes) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<uint8_t> dist(0, 255);
    std::string out;
    out.reserve(bytes * 2);
    static constexpr char hex[] = "0123456789abcdef";
    for (size_t i = 0; i < bytes; ++i) {
        uint8_t b = dist(gen);
        out += hex[b >> 4];
        out += hex[b & 0xf];
    }
    return out;
}

// Build iLink auth headers (shared by all API calls).
std::vector<std::string> ilink_headers(const std::string& token, size_t body_len) {
    return {
        "Content-Type: application/json",
        std::string("Authorization: Bearer ") + token,
        "AuthorizationType: ilink_bot_token",
        std::string("Content-Length: ") + std::to_string(body_len),
    };
}

// Parse a single item from item_list; returns ILinkItem.
ILinkItem parse_item(const json& j) {
    ILinkItem item;
    item.type = j.value("type", 0);
    if (item.type == 1) {  // MSG_ITEM_TEXT
        item.text = j.value(json::json_pointer("/text_item/text"), std::string{});
    } else if (item.type == 3) {  // MSG_ITEM_VOICE (iLink auto-transcribes)
        item.text = j.value(json::json_pointer("/voice_item/text"), std::string{});
    }
    return item;
}

} // anonymous namespace

// ── credential persistence ─────────────────────────────────────────────────

ILinkCreds load_creds(const Config& cfg) {
    if (!fs::exists(cfg.cred_path)) {
        throw std::runtime_error(
            "No WeChat credentials at " + cfg.cred_path +
            ". Run: ./wcp-setup");
    }
    std::ifstream f(cfg.cred_path);
    auto j = json::parse(f);
    ILinkCreds creds;
    creds.token    = j.at("token").get<std::string>();
    creds.base_url = j.value("baseUrl", "https://ilinkai.weixin.qq.com");
    return creds;
}

std::string load_sync_buf(const Config& cfg) {
    std::ifstream f(cfg.sync_path);
    if (!f) return {};
    return std::string(std::istreambuf_iterator<char>(f), {});
}

void save_sync_buf(const Config& cfg, std::string_view buf) {
    fs::create_directories(fs::path(cfg.sync_path).parent_path());
    std::ofstream f(cfg.sync_path, std::ios::trunc);
    f << buf;
}

// ── get_updates ────────────────────────────────────────────────────────────

Result<GetUpdatesResponse> get_updates(
    const ILinkCreds& creds,
    std::string_view  sync_buf,
    const Config&     cfg)
{
    json payload;
    payload["get_updates_buf"] = std::string(sync_buf);
    std::string body = payload.dump();

    auto res = http_post(
        creds.base_url + "/ilink/bot/getupdates",
        ilink_headers(creds.token, body.size()),
        body,
        cfg.long_poll_ms);

    if (!res.ok) {
        // Timeout maps to CURLE_OPERATION_TIMEDOUT — treat as empty poll
        if (res.error.find("Timeout") != std::string::npos ||
            res.error.find("timed out") != std::string::npos) {
            return Result<GetUpdatesResponse>::success({0, {}, std::string(sync_buf)});
        }
        return Result<GetUpdatesResponse>::failure(std::move(res.error));
    }

    auto j = json::parse(res.value.body, nullptr, false);
    if (j.is_discarded())
        return Result<GetUpdatesResponse>::failure("iLink: invalid JSON response");

    GetUpdatesResponse out;
    out.errcode  = j.value("errcode", 0);
    out.next_buf = j.value("get_updates_buf", std::string(sync_buf));

    for (const auto& raw : j.value("msgs", json::array())) {
        ILinkMessage msg;
        msg.message_type  = raw.value("message_type", 0);
        msg.from_user_id  = raw.value("from_user_id",  "");
        msg.context_token = raw.value("context_token", "");
        for (const auto& item : raw.value("item_list", json::array()))
            msg.items.push_back(parse_item(item));
        out.msgs.push_back(std::move(msg));
    }

    return Result<GetUpdatesResponse>::success(std::move(out));
}

// ── send_reply ─────────────────────────────────────────────────────────────

Result<void> send_reply(
    const ILinkCreds& creds,
    std::string_view  to_user,
    std::string_view  text,
    std::string_view  context_token,
    const Config&     cfg)
{
    std::string client_id = "wcp-" + random_hex(8);

    json payload;
    payload["msg"] = {
        {"from_user_id",  ""},
        {"to_user_id",    std::string(to_user)},
        {"client_id",     client_id},
        {"message_type",  2},
        {"message_state", 2},
        {"context_token", std::string(context_token)},
        {"item_list", json::array({
            {{"type", 1}, {"text_item", {{"text", std::string(text)}}}}
        })},
    };
    std::string body = payload.dump();

    auto res = http_post(
        creds.base_url + "/ilink/bot/sendmessage",
        ilink_headers(creds.token, body.size()),
        body,
        cfg.api_timeout_ms);

    if (!res.ok) return Result<void>::failure(std::move(res.error));
    if (res.value.status_code != 200)
        return Result<void>::failure(
            "sendMessage HTTP " + std::to_string(res.value.status_code) +
            ": " + res.value.body);

    return Result<void>::success();
}

// ── QR login helpers ───────────────────────────────────────────────────────

Result<QRCodeInfo> get_bot_qrcode(std::string_view base_url, const Config& cfg) {
    std::string body = R"({"bot_type":3})";
    auto res = http_post(
        std::string(base_url) + "/ilink/bot/get_bot_qrcode",
        {"Content-Type: application/json"},
        body,
        cfg.api_timeout_ms);

    if (!res.ok) return Result<QRCodeInfo>::failure(std::move(res.error));

    auto j = json::parse(res.value.body, nullptr, false);
    if (j.is_discarded()) return Result<QRCodeInfo>::failure("Invalid JSON");

    QRCodeInfo info;
    info.qr_url = j.value("qr_url", "");
    info.ticket = j.value("ticket", "");
    info.token  = j.value("token",  "");
    return Result<QRCodeInfo>::success(std::move(info));
}

Result<QRLoginStatus> poll_qr_login(
    std::string_view base_url,
    std::string_view ticket,
    const Config&    cfg)
{
    json payload;
    payload["ticket"] = std::string(ticket);
    std::string body = payload.dump();

    auto res = http_post(
        std::string(base_url) + "/ilink/bot/query_qrcode_status",
        {"Content-Type: application/json"},
        body,
        cfg.api_timeout_ms);

    if (!res.ok) return Result<QRLoginStatus>::failure(std::move(res.error));

    auto j = json::parse(res.value.body, nullptr, false);
    if (j.is_discarded()) return Result<QRLoginStatus>::failure("Invalid JSON");

    QRLoginStatus s;
    s.status = j.value("status", 0);
    if (j.contains("token") && !j["token"].is_null())
        s.token = j["token"].get<std::string>();
    return Result<QRLoginStatus>::success(std::move(s));
}

} // namespace wcp
